File: python/17_LetterCombinations.py
from unittest import TestCase

# https://leetcode.com/problems/letter-combinations-of-a-phone-number/


class LetterCombinations(TestCase):
    LETTERS = [
        "abc",
        "def",
        "ghi",
        "jkl",
        "mno",
        "pqrs",
        "tuv",
        "wxyz",
    ]

    def letterCombinations(self, digits):
        result = []

        def build(prefix, i):
            if i >= len(digits):
                result.append(prefix)
            else:
                d = ord(digits[i]) - ord('2')
                for ch in self.LETTERS[d]:
                    build(prefix + ch, i + 1)

        if digits:
            build("", 0)
        return result


    def test1(self):
        self.assertEqual(["ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"], self.letterCombinations("23"))

    def test2(self):
        self.assertEqual([], self.letterCombinations(""))

File: kotlin/src/main/kotlin/17.LetterCombinations.kt
import org.junit.Assert
import org.junit.Test

// https://leetcode.com/problems/letter-combinations-of-a-phone-number/

val LETTERS = listOf(
    "abc",
    "def",
    "ghi",
    "jkl",
    "mno",
    "pqrs",
    "tuv",
    "wxyz"
)

class LetterCombinations{
  fun letterCombinations(digits: String): List<String> {
    val result = mutableListOf<String>()

    fun build(prefix: String, i: Int){
      if(i == digits.length)
        result.add(prefix)
      else{
        val j = digits[i].toInt() - '2'.toInt()
        for(c in LETTERS[j])
          build(prefix+c, i+1)
      }
    }

    if(digits.isNotEmpty())
      build("", 0)
    return result
  }


  @Test
  fun test1(){
    Assert.assertEquals(listOf("ad", "ae", "af", "bd", "be", "bf", "cd", "ce", "cf"), letterCombinations("23"))
  }

  @Test
  fun blank(){
    Assert.assertEquals(listOf<String>(), letterCombinations(""))
  }
}
